lib/atscppapi/examples/websocket/WebSocket.cc

```diff
@@ -0,0 +1,311 @@
+#define WS_RSV3 0x10
+#define WS_MASKED 0x80
+#define WS_OPCODE 0x0F
+#define WS_FIN 0x80
+#define WS_LENGTH 0x7F
+#define WS_16BIT_LEN 126
+#define WS_64BIT_LEN 127
+
+void TSPluginInit(int argc, const char* argv[])
+{
+    RegisterGlobalPlugin("WebSocket", "Apache", "support@example.com");
+    new WebSocketInstaller();
+}
+
+// WebSocketInstaller
+
+WebSocketInstaller::WebSocketInstaller()
+    : GlobalPlugin(true /* ignore internal transactions */)
+{
+    GlobalPlugin::registerHook(Plugin::HOOK_READ_REQUEST_HEADERS_PRE_REMAP);
+}
+
+#define WS_DIGEST_MAX ATS_BASE64_ENCODE_DSTLEN(20)
+static const std::string magic = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+
+static std::string ws_digest(std::string const& key) {
+    EVP_MD_CTX digest;
+    EVP_MD_CTX_init(&digest);
+
+    if (!EVP_DigestInit_ex(&digest, EVP_sha1(), NULL)) {
+        EVP_MD_CTX_cleanup(&digest);
+        return "init-failed";
+    }
+    if (!EVP_DigestUpdate(&digest, key.data(), key.length())) {
+        EVP_MD_CTX_cleanup(&digest);
+        return "update1-failed";
+    }
+    if (!EVP_DigestUpdate(&digest, magic.data(), magic.length())) {
+        EVP_MD_CTX_cleanup(&digest);
+        return "update2-failed";
+    }
+
+    unsigned char hash_buf[EVP_MAX_MD_SIZE];
+    unsigned int hash_len = 0;
+    if (!EVP_DigestFinal_ex(&digest, hash_buf, &hash_len)) {
+        EVP_MD_CTX_cleanup(&digest);
+        return "final-failed";
+    }
+    EVP_MD_CTX_cleanup(&digest);
+    if (hash_len != 20) {
+        return "bad-hash-length";
+    }
+
+    char digest_buf[WS_DIGEST_MAX];
+    size_t digest_len = 0;
+
+    ats_base64_encode(hash_buf, hash_len, digest_buf, WS_DIGEST_MAX, &digest_len);
+
+    return std::string((char*)digest_buf, digest_len);
+}
+
+void WebSocketInstaller::handleReadRequestHeadersPreRemap(Transaction &transaction)
+{
+    SAY("Incoming request.");
+    transaction.addPlugin(new WebSocket(transaction));
+    transaction.resume();
+}
+
+// WebSocket implementation.
+
+WebSocket::WebSocket(Transaction& transaction)
+    : InterceptPlugin(transaction, InterceptPlugin::SERVER_INTERCEPT)
+    , pos_(0)
+    , ctrl_(0)
+    , mask_len_(0)
+    , msg_len_(0)
+    , ws_handshake_done_(false)
+{
+    if (isWebsocket()) {
+        ws_key_ = transaction.getClientRequest().getHeaders().values("sec-websocket-key");
+    }
+}
+
+WebSocket::~WebSocket()
+{
+    SAY("WebSocket finished.");
+}
+
+void WebSocket::ws_send(std::string const& data, int code)
+{
+    size_t len = data.length();
+
+    std::string frame;
+    frame.reserve(len + 10);
+    frame += char(code);
+
+    int len_len;
+    if (len <= 125) {
+        frame += char(len);
+        len_len = 0;
+    } else if (len <= UINT16_MAX) {
+        frame += char(WS_16BIT_LEN);
+        len_len = 2;
+    } else {
+        frame += char(WS_64BIT_LEN);
+        len_len = 8;
+    }
+    // Convert length to big-endian bytes.
+    while (--len_len >= 0) {
+        frame += char((len >> (8*len_len)) & 0xFF);
```

Comment:
To convert the length to big endian just use:
htons() or htobe64(), same idea as below.